ratbag_emu/device.py

```diff
@@ -0,0 +1,272 @@
+        '''
+        Generates a random name
+        '''
+        device_names = [
+                'mara', 'capybara', 'porcupine', 'paca',
+                'vole', 'woodrat', 'gerbil', 'shrew',
+                'hutia', 'beaver', 'squirrel', 'chinchilla',
+                'rabbit', 'viscacha', 'hare', 'degu',
+                'gundi', 'acouchy', 'nutria', 'paca',
+                'hamster', 'zokor', 'chipmunk', 'gopher',
+                'marmot', 'groundhog', 'suslik', 'agouti',
+                'blesmol',
+        ]
+
+        device_attr = [
+                'sobbing', 'whooping', 'barking', 'yapping',
+                'howling', 'squawking', 'cheering', 'warbling',
+                'thundering', 'booming', 'blustering', 'humming',
+                'crying', 'bawling', 'roaring', 'raging',
+                'chanting', 'crooning', 'murmuring', 'bellowing',
+                'wailing', 'weeping', 'screaming', 'yelling',
+                'yodeling', 'singing', 'honking', 'hooting',
+                'whispering', 'hollering',
+        ]
+
+        name = device_names[random.randint(0, len(device_names)-1)]
+        attr = device_attr[random.randint(0, len(device_attr)-1)]
+        return '-'.join([attr, name])
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def info(self):
+        return self._info
+
+    @property
+    def rdescs(self):
+        return self._rdescs
+
+    @property
+    def actuators(self):
+        return self._actuators
+
+    @actuators.setter
+    def actuators(self, val):
+        # Make sure we don't have actuators which will act on the same keys
+        seen = []
+        for keys in [a.keys for a in val]:
+            for el in keys:
+                assert el not in seen
+                seen.append(el)
+
+        self._actuators = val
+
+    def destroy(self):
+        for endpoint in self.endpoints:
+            endpoint.destroy()
+
+    def transform_action(self, data: Dict[str, Any]):
+        '''
+        Transforms high-level action according to the actuators
+
+        A high-level action will have the x, y values in mm. This values will
+        be converted to pixel by the device actuators (in this case, the
+        sensor/dpi actuator)
+
+        :param action:  high-level action
+        '''
+        hid_data: Dict[str, Any] = {}
+
+        for actuator in self.actuators:
+            hid_data.update(actuator.transform(data.copy()))
+
+        return hid_data
+
+    def send_hid_action(self, action: object):
+        '''
+        Sends a HID action
+
+        We assume there's only one endpoint for each type of action (mouse,
+        keyboard, button, etc.) so we send the action to all endpoints. The
+        endpoint will only act on the action if it supports it.
+
+        :param action:  HID action
+        '''
+        for endpoint in self.endpoints:
+            # FIXME: global data (0x11)?
+            endpoint.send(endpoint.create_report(action, 0x11))
```

Comment:
What is `0x11`?

Reply:
I am not sure but it was in the original code and I didn't remove. I am going to see if it is actually required, I was under the impression that it is.

Reply:
It isn't. Removed.